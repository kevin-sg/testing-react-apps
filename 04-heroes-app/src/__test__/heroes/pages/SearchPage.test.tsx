import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";

import { SearchPage } from "@/heroes";

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("Test in <SearchPage/>", () => {
  beforeEach(() => jest.clearAllMocks());

  test("should show correctly and return default values", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  test("should show input value of the query", () => {
    const mockedQuery = "batman";
    render(
      <MemoryRouter initialEntries={[`/search?q=${mockedQuery}`]}>
        <SearchPage />
      </MemoryRouter>,
    );

    const divElement = screen.getByTestId("test-alert");
    const inputElement = screen.getByTestId("test-input-value");
    const imgElement = screen.getByRole("img").getAttribute("src");

    expect(divElement.style.display).toBe("none");
    expect(divElement.className).toContain("alert");
    expect(inputElement.getAttribute("value")).toBe(mockedQuery);
    expect(imgElement).toContain("src/assets/heroes/dc-batman.jpg");
  });

  test("should show an error if the hero is not found (batman123)", () => {
    const mockedQuery = "batman123";
    render(
      <MemoryRouter initialEntries={[`/search?q=${mockedQuery}`]}>
        <SearchPage />
      </MemoryRouter>,
    );

    const divElement = screen.getByTestId("test-alert");

    expect(divElement.style.display).toBeFalsy();
    expect(divElement.textContent).toBe(`No hero with ${mockedQuery}`);
  });

  test("should call navigate to the new screen", async () => {
    const mockedQuery = "flash";
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>,
    );

    const inputElement = screen.getByTestId("test-input-value");
    const buttonElement = screen.getByTestId("on-submit-click");

    await userEvent.type(inputElement, mockedQuery);
    await userEvent.click(buttonElement);

    expect(mockedUsedNavigate).toHaveBeenCalled();
    expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUsedNavigate).toHaveBeenCalledWith(`?q=${mockedQuery}`);
  });
});
