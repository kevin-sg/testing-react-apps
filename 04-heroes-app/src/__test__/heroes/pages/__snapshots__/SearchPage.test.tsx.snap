// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <SearchPage/> should show correctly and return default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          data-testid="test-input-value"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          data-testid="on-submit-click"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        data-testid="test-alert"
        role="alert"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
